KAFKA-12170: Fix for Connect Cast SMT to correctly transform a Byte array into a string

sknop opened the pull request:
Cast SMT transformation for bytes -> string.
Without this fix, the conversion becomes

ByteBuffer.toString(), which always gives one of these useless results:
    "java.nio.HeapByteBuffer[pos=0 lim=4 cap=4]" 
    "[B@5ec0a365" (for byte array)
    

With this change, the byte array is converted into a hex string of the
byte buffer content, for example

    "FEDCBA9876543210"

Completed with test case and successfully tried out in a
real database conversion.


### Committer Checklist (excluded from commit message)
- [X ] Verify design and implementation 
- [ X] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

sknop committed:
Fix for https://issues.apache.org/jira/browse/KAFKA-12170

Cast SMT transformation for bytes -> string.
Without this fix, the conversion becomes

ByteBuffer.toString(), which always gives this useless result:
    "java.nio.HeapByteBuffer[pos=0 lim=4 cap=4]"

With this change, the byte array is converted into a hex string of the
byte buffer content, for example

    "FEDCBA9876543210"

Completed with test case and successfully tried out in a
real database conversion.

sknop committed:
Fix cast byte array -> string for byte[]

It turns out that not all connectors produce a HeapBuffer to store bytes.
SAP Hana, specifically, just stores a simple byte array that was not
captured by the cast transformer.

Completed with test case.

Fix for https://issues.apache.org/jira/browse/KAFKA-12170

g1geordie commented:
@sknop  hello , It seems like there are some style error .
You can use `./gradlew checkstyleMain checkstyleTest` to cehck .

sknop committed:
Fixes after failed style check

sknop commented:
> @sknop hello , It seems like there are some style error .
> You can use `./gradlew checkstyleMain checkstyleTest` to cehck .

Thank you for your feedback. I have adjusted the code and it now passes the style check.

mimaison commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -360,15 +361,30 @@ else if (value instanceof String)
             java.util.Date dateValue = (java.util.Date) value;
             return Values.dateFormatFor(dateValue).format(dateValue);
+        } else if (value instanceof ByteBuffer) {
+            ByteBuffer byteBuffer = (ByteBuffer) value;
```
We can inline this like you've done for the byte[] case below

mimaison commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -360,15 +361,30 @@ else if (value instanceof String)
+            ByteBuffer byteBuffer = (ByteBuffer) value;
+
+            return castByteArrayToString(byteBuffer.array());
+        } else if (value.getClass() == byte[].class) {
```
Let's use `value instanceof byte[]` so it's like the other cases

mimaison commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -360,15 +361,30 @@ else if (value instanceof String)
             throw new DataException("Unexpected type in Cast transformation: " + value.getClass());
     }
 
+
```
We don't need this extra blank line

mimaison commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/CastTest.java:
```diff
@@ -425,7 +426,21 @@ public void castLogicalToString() {
+        byteBuffer.put((byte) 0x10);
+        byteBuffer.flip();
+
+        byte[] byteArray = Arrays.copyOf(byteBuffer.array(), byteBuffer.array().length);
```
What about doing something like:
```java
byte[] byteArray = new byte[] {(byte) 0xFE, (byte) 0xDC, (byte) 0xBA, (byte) 0x98, 0x76, 0x54, 0x32, 0x10};
ByteBuffer byteBuffer = ByteBuffer.wrap(Arrays.copyOf(byteArray, byteArray.length));
```

mimaison reviewed (changes requested):
Thanks for the PR!
It looks good overall, I just left a few minor comments

sknop committed:
Changes made after code review by Mickael Maison (@mimaison)

sknop commented:
Hi Mickael

Thank you so much for your code review. I have adjusted the code according to your suggestions, they all make sense to me.

Cheers

Sven Erik
seknop@gmail.com


On 18 Feb 2021, 16:40 +0000, apache/kafka <reply@reply.github.com>, wrote:
>
> @mimaison

kkonstantine commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -364,11 +365,24 @@ private static String castToString(Object value) {
             return Values.dateFormatFor(dateValue).format(dateValue);
+        } else if (value instanceof ByteBuffer) {
+            ByteBuffer byteBuffer = (ByteBuffer) value;
+            return castByteArrayToString(byteBuffer.array());
```
formally we should almost always check that a `ByteBuffer` is indeed backed by an array with `hasArray` or this part my throw exceptions now: https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#array--

sknop replied:
That is a good point, but what should happen in that case? 

I assume we start with hasArray() and only process the content if an array is available, as I understand it? ByteBuffer.remaining() would contain the number of bytes available ... I'll go and write some test cases to understand this better.

sknop replied:
I added code that checks for the existence of the array and otherwise, I extract the content through ByteBuffer.get(byte[] array).

kkonstantine commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -364,11 +365,24 @@ private static String castToString(Object value) {
+    private static String castByteArrayToString(byte[] array) {
+        StringBuilder sbuf = new StringBuilder();
+        for (byte b : array) {
+            sbuf.append(String.format("%02X", b));
```
Also, I'm a bit curious. Why are we selecting hex here and not base64 for example?
Hex of course is less efficient. 
https://issues.apache.org/jira/browse/KAFKA-12170 does not mention anything about the rational, so I thought I'd ask. 

@rhauch wdyt about the representation here?

sknop replied:
The reason I suggested Hex representation was that in the use case I encountered - database PK field encoded as BINARY retrieved via CDC Source Connector - the representation in the SQL CLI was Hex, so choosing the same format made it easy to compare source and target.

I would suggest that base64 encoding would validate its own dedicated transformer, with an ability to choose padding, for example

rhauch replied:
As noted in my previous comment, I agree with @kkonstantine that base64 would be preferable. Doing that would align better with the existing `Values` class used in Connect's header converter mechanism.

If we want to add support for multiple encodings, we would need to have a KIP since it would likely mean changing the SMT configuration.

sknop replied:
If base64 is the better choice we shall go with that. Would we need to adjust the JIRA ticket to reflect this?

I just ran a little test in the JShell:

jshell> byte[] byteArray = new byte[] {(byte) 0xFE, (byte) 0xDC, (byte) 0xBA, (byte) 0x98, 0x76, 0x54, 0x32, 0x10};
byteArray ==> byte[8] { -2, -36, -70, -104, 118, 84, 50, 16 }

jshell> Base64.getEncoder().encode(byteArray)
$2 ==> byte[12] { 47, 116, 121, 54, 109, 72, 90, 85, 77, 104, 65, 61 }

jshell> Base64.getEncoder().encodeToString(byteArray)
$3 ==> "/ty6mHZUMhA="

jshell> Base64.getEncoder().withoutPadding().encodeToString(byteArray)
$4 ==> "/ty6mHZUMhA"

Would you suggest using padding or not?

rhauch replied:
Connect's `Values` code just uses `Base64.getEncoder().encodeToString((byte[]) value)` and does not disable padding. Seems like we should follow the same pattern here. Thoughts, @kkonstantine ?

kkonstantine replied:
Yes, I agree. `encodeToString` is the method I had in mind. Used in `Values` and JsonConverter for a conversion of bytes to string. It's also more efficient.

kkonstantine reviewed (commented):
Thanks for the PR @sknop 
Left a few comments inline.

sknop committed:
Address code review feedback: protect case where ByteBuffer has not array.

If the ByteBuffer does not have a backing array, we use get(byte[])
to extract the content.

rhauch commented:
Before this fix, the changes made in #4820 ([KAFKA-6684](https://issues.apache.org/jira/browse/KAFKA-6684)) resulted in `toString()` being called on `byte[]` and `ByteBuffer`. As highlighted in this PR and issue that `ByteBuffer.toString()` is not useful, but the `toString()` on `byte[]` still works. This PR seems to change that behavior, which would not be backward compatible.

The discussion on PR #4820 also talked about making this compatible with `Values.convertToString(...)`, which for `byte[]` and `ByteBuffer` results in a base 64 encoded string (with `ISO-8859-1` encoding). See the [Values code for details](https://github.com/apache/kafka/blob/95f51539c8d0b88bd7f285011d42e2d1117107de/connect/api/src/main/java/org/apache/kafka/connect/data/Values.java#L673-L682).

Because of this, it seems much more sensible to also use base64 here for `ByteBuffer` so it matches the existing behavior with `byte[]`. 

I agree that it maybe doesn't suffice in all user situations, so if we also want to support other encodings we'd need other config changes that will require using the KIP mechanism to propose such enhancements.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -364,11 +365,31 @@ private static String castToString(Object value) {
             return Values.dateFormatFor(dateValue).format(dateValue);
+        } else if (value instanceof ByteBuffer) {
+            ByteBuffer byteBuffer = (ByteBuffer) value;
+            if (byteBuffer.hasArray()) {
+                return castByteArrayToString(byteBuffer.array());
+            }
+            else {
+                byte[] array = new byte[byteBuffer.remaining()];
+                byteBuffer.get(array);
+                return castByteArrayToString(array);
+            }
```
The use of `ByteBuffer.get(...)` here does not account for the fact that it may not be positioned at the beginning. Kafka has two `Utils.readBytes(...)` methods that we should probably use. This code would then simplify to:
```suggestion
            byte[] rawBytes = Utils.readBytes(byteBuffer);
            return castByteArrayToString(rawBytes);
```

sknop replied:
That sounds like an excellent plan, I did not know this class existed. Changed.

sknop replied:
This sounds like an excellent plan and passes the tests.

sknop committed:
Kafka has ready-made Utils class for extracting bytes from a ByteBuffer.

Change made after code review.

kkonstantine reviewed (commented):
Thanks for the detailed comments @rhauch 
@sknop I agree with @rhauch that base64 would be preferable and consistent with our previous choices. I also agree to punt support of multiple encodings for now.

sknop committed:
Changed conversion to a Base64-encoded string instead of Hex

This matches the internal format used in other parts of the Connect code.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -39,12 +40,8 @@
-import java.util.Map;
-import java.util.Set;
+import java.nio.ByteBuffer;
+import java.util.*;
```
Wildcard imports should be caught by checkstyle, and should fail the build. In any case, please replace with non-wildcard imports.

sknop replied:
IntelliJ did this automatically for me - I have changed the code accordingly.

rhauch reviewed (commented):
One import-related fix, but otherwise looks good.

sknop committed:
Replaced wildcard with single class imports.

Change requested by code review.

sknop commented:
What about documentation? 

Where is the documentation stored and what is the process to change it? It might be useful to point out that the Cast SMT will create base64 strings for byte arrays.

rhauch commented:
The documentation is auto-generated from the code, via [TransformationDoc](https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/tools/TransformationDoc.java).

sknop commented:
Thank you @rhauch. Just to make sure I understand this correctly - the types mentioned in the cast string are the _target_ casts, not the source tasks. The current code does not allow casting to a byte array, only casting _from_ a byte array.
As I understand it, therefore no documentation changes are required (except maybe for release notes)?

rhauch commented:
If we don't clarify the documentation, then I think users will be very confused. Can you take a stab at improving/expanding the documentation a bit to clarify the input and output types?

sknop commented:
Yes of course, totally agree.

sknop committed:
Updated Overview and Config after code review

Thank you @rhauch for your feedback.

rhauch reviewed (approved):
Thanks, @sknop! Looks great.

@mimaison, you have a "request changes", so can you please review? Thanks!

mimaison reviewed (approved):
Thanks for the PR. LGTM

kkonstantine reviewed (approved):
This LGTM
Thanks @sknop

mimaison commented:
Test failures are not related:
```
JDK 8 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationSSLTest.testOneWayReplicationWithAutoOffsetSync()
JDK 11 / org.apache.kafka.connect.integration.RebalanceSourceConnectorsIntegrationTest.testMultipleWorkersRejoining
```

Merging to trunk
